File: Sliding_Window/239.Sliding_Window_Maximum/Python/239.Sliding_Window_Maximum.py
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        #==============================#
        # Deque-based traversal method #
        #==============================#

        ############
        #Initialize
        ##### Length of nums array #####
        len_nums = len(nums)

        ##### Record deque #####
        record_dq = []

        ##### Result array #####
        res_arry = []


        ############################
        #Deque-based loop traversal
        ##### Step 1: Looped-traversal for the first fixed-window #####
        for nums_idx in range(k):
            while (record_dq and ((record_dq[(-1)])[0] <= nums[nums_idx])):
                record_dq.pop() #Keep updating/popped

            record_dq.append((nums[nums_idx], nums_idx)) #Keep updating/recording

        ##### Step 2: Looped-traversal for the remainings #####
        res_arry.append((record_dq[0])[0]) #Update/Record
        
        for nums_idx in range(k, len_nums):
            while (record_dq and ((record_dq[0])[1] <= (nums_idx - k))):
                record_dq.pop(0) #Keep updating/popped

            while (record_dq and ((record_dq[(-1)])[0] <= nums[nums_idx])):
                record_dq.pop() #Keep updating/popped

            record_dq.append((nums[nums_idx], nums_idx)) #Keep updating/recording

            res_arry.append((record_dq[0])[0]) #Keep updating/recording

        return res_arry

File: Sliding_Window/239.Sliding_Window_Maximum/C++/239.Sliding_Window_Maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int> & nums, int k) 
    {
        //====================================//
        // Input type:                        //
        //   - vector<int> ref. (nums vector) //
        //   - int (k-th size)                //
        // Return type:                       //
        //   - vector<int> (result vector)    //
        //====================================//

        /*Initialize*/
        ///// Length of nums vector /////
        int len_nums = nums.size();

        ///// Record deque /////
        deque<pair<int, int>> record_dq;

        ///// Result vector /////
        vector<int> res_vect;


        /*Dequeue-based loop traversal*/
        ///// Step 1: Looped-traversal for the first fixed-window /////
        for (int nums_idx = 0; (nums_idx < k); (nums_idx++)) //K-th
        {
            while ((! record_dq.empty()) && ((record_dq.back()).first <= nums[nums_idx])) { record_dq.pop_back(); } //Keep updating, popped
                
            record_dq.emplace_back(nums[nums_idx], nums_idx); //Keep updating, recording

        } //K-th

        ///// Step 2: Looped-traversal for the remainings /////
        res_vect.emplace_back((record_dq.front()).first); //Update, Record

        for (int nums_idx = k; (nums_idx < len_nums); (nums_idx++)) //Remains
        {
            while ((! record_dq.empty()) && ((record_dq.front()).second <= (nums_idx - k))) { record_dq.pop_front(); } //Keep updating, popped

            while ((! record_dq.empty()) && ((record_dq.back()).first <= nums[nums_idx])) { record_dq.pop_back(); } //Keep updating, popped

            record_dq.emplace_back(nums[nums_idx], nums_idx); //Keep updating, recording

            res_vect.emplace_back((record_dq.front()).first); //Keep updating, recording

        } //Remains

        return res_vect;
    }
};
